src/libdredd/src/mutation_replace_unary_operator.cc

```diff
@@ -0,0 +1,263 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "libdredd/mutation_replace_unary_operator.h"
+
+#include <cassert>
+#include <sstream>
+#include <vector>
+
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/Expr.h"
+#include "clang/AST/OperationKinds.h"
+#include "clang/AST/Type.h"
+#include "clang/Basic/SourceLocation.h"
+#include "clang/Lex/Preprocessor.h"
+#include "clang/Rewrite/Core/Rewriter.h"
+#include "libdredd/util.h"
+#include "llvm/ADT/StringRef.h"
+
+namespace dredd {
+
+MutationReplaceUnaryOperator::MutationReplaceUnaryOperator(
+    const clang::UnaryOperator& unary_operator)
+    : unary_operator_(unary_operator) {}
+
+bool MutationReplaceUnaryOperator::IsPrefix(clang::UnaryOperatorKind op) {
+  return op != clang::UO_PostInc && op != clang::UO_PostDec;
+}
+
+bool MutationReplaceUnaryOperator::IsValidReplacementOperator(
+    clang::UnaryOperatorKind op) const {
+  if (!unary_operator_.getSubExpr()->isLValue() &&
+      (op == clang::UO_PreInc || op == clang::UO_PreDec ||
+       op == clang::UO_PostInc || op == clang::UO_PostDec)) {
+    return false;
+  }
+
+  if (unary_operator_.isLValue() &&
+      !(op == clang::UO_PreInc || op == clang::UO_PreDec)) {
+    return false;
+  }
+
+  if (op == clang::UO_Not && unary_operator_.getSubExpr()
+                                 ->getType()
+                                 ->getAs<clang::BuiltinType>()
+                                 ->isFloatingPoint()) {
+    return false;
+  }
+
+  return true;
+}
+
+std::string MutationReplaceUnaryOperator::GetFunctionName(
+    clang::ASTContext& ast_context) const {
+  std::string result = "__dredd_replace_unary_operator_";
+
+  // A string corresponding to the unary operator forms part of the name of the
+  // mutation function, to differentiate mutation functions for different
+  // operators
+  switch (unary_operator_.getOpcode()) {
+    case clang::UnaryOperatorKind::UO_Plus:
+      result += "Plus";
+      break;
+    case clang::UnaryOperatorKind::UO_Minus:
+      result += "Minus";
+      break;
+    case clang::UnaryOperatorKind::UO_Not:
+      result += "Not";
+      break;
+    case clang::UnaryOperatorKind::UO_PreDec:
+      result += "PreDec";
+      break;
+    case clang::UnaryOperatorKind::UO_PostDec:
+      result += "PostDec";
+      break;
+    case clang::UnaryOperatorKind::UO_PreInc:
+      result += "PreInc";
+      break;
+    case clang::UnaryOperatorKind::UO_PostInc:
+      result += "PostInc";
+      break;
+    case clang::UnaryOperatorKind::UO_LNot:
+      result += "LNot";
+      break;
+    default:
+      assert(false && "Unsupported opcode");
+  }
+
+  std::string input_qualifier;
```

Comment:
Get rid of this (suggestion below on what to do instead)

Reply:
Done.